src/openpa/src/primitives/opa_gcc_intrinsics.h

```diff
@@ -34,54 +42,86 @@ typedef struct {
    may not be true at all. */
 static inline int OPA_load_int(const OPA_int_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_RELAXED);
+#else
     return ptr->v;
+#endif
 }
 
 static inline void OPA_store_int(OPA_int_t * ptr, int val)
 {
+#ifdef USE_GCC_ATOMIC
+    __atomic_store_n(&ptr->v, val, __ATOMIC_RELAXED);
+#else
     ptr->v = val;
+#endif
 }
 
 static inline void *OPA_load_ptr(const OPA_ptr_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_RELAXED);
+#else
     return ptr->v;
+#endif
 }
 
 static inline void OPA_store_ptr(OPA_ptr_t * ptr, void *val)
 {
+#ifdef USE_GCC_ATOMIC
+    __atomic_store_n(&ptr->v, val, __ATOMIC_RELAXED);
+#else
     ptr->v = val;
+#endif
 }
 
 static inline int OPA_load_acquire_int(const OPA_int_t * ptr)
 {
+#ifdef USE_GCC_ATOMIC
+    return __atomic_load_n(&ptr->v, __ATOMIC_ACQUIRE);
+#else
     volatile int i = 0;
     int tmp;
-    tmp = ptr->v;
     __sync_lock_test_and_set(&i, 1);    /* guarantees acquire semantics */
+    tmp = ptr->v;
     return tmp;
+#endif
 }
 
 static inline void OPA_store_release_int(OPA_int_t * ptr, int val)
 {
+#ifdef USE_GCC_ATOMIC
+    __atomic_store_n(&ptr->v, val, __ATOMIC_RELEASE);
+#else
     volatile int i = 1;
-    __sync_lock_release(&i);    /* guarantees release semantics */
     ptr->v = val;
+    __sync_lock_release(&i);    /* guarantees release semantics */
```

Comment:
For the same reason as `OPA_load_acquire_int()`, I don't think this reordering is okay.

Reply:
You are correct this is wrong, but, as commented above, with the `__sync` builtins, this has to resolve to a full barrier.  Races will happen with code inside the critical section regardless of which side `ptr->v = val;` is placed on because Release semantics allow younger stores in program order to cross the barrier and order before older memory accesses, unless the store we are trying to order is itself tagged as a store-release.

For Arm at least, refer to page B2-95 for the conditions a R or W is ordered before a barrier.